File: app/channels/telegram/keyboards.py
"""Telegram reply keyboards."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, KeyboardButton, ReplyKeyboardMarkup, ReplyKeyboardRemove

from app.domain.telegram_menu import main_menu_button_texts

PHONE_BUTTON_TEXT = "📱 Telefon raqamni yuborish"

_PHONE_BUTTON: Dict[str, str] = {
    "uz_lat": "📱 Telefon raqamni yuborish",
    "uz_cyrl": "📱 Телефон рақамни юбориш",
    "ru": "📱 Отправить номер телефона",
    "en": "📱 Send phone number",
    "zh": "📱 发送电话号码",
}


def phone_request_keyboard(lang: str = "uz_lat") -> ReplyKeyboardMarkup:
    label = _PHONE_BUTTON.get(lang) or _PHONE_BUTTON.get("uz_lat", PHONE_BUTTON_TEXT)
    return ReplyKeyboardMarkup(
        [[KeyboardButton(label, request_contact=True)]],
        resize_keyboard=True,
        one_time_keyboard=True,
    )


def remove_keyboard() -> ReplyKeyboardRemove:
    return ReplyKeyboardRemove()


def main_menu_keyboard(lang: str = "uz_lat") -> ReplyKeyboardMarkup:
    """Asosiy menyu: 2×2 + mahsulot qidirish."""
    labels = main_menu_button_texts(lang)
    return ReplyKeyboardMarkup(
        [
            [
                KeyboardButton(labels["callback"]),
                KeyboardButton(labels["new_chat"]),
            ],
            [
                KeyboardButton(labels["language"]),
                KeyboardButton(labels["help"]),
            ],
            [KeyboardButton(labels["product_search"])],
        ],
        resize_keyboard=True,
        is_persistent=True,
    )


def inline_keyboard_from_extra(extra: Optional[Dict[str, Any]]) -> Optional[InlineKeyboardMarkup]:
    if not extra:
        return None
    rows = extra.get("inline_keyboard")
    if not isinstance(rows, list) or not rows:
        return None
    markup_rows: List[List[InlineKeyboardButton]] = []
    for row in rows:
        if not isinstance(row, list):
            continue
        buttons = []
        for btn in row:
            if not isinstance(btn, dict) or not btn.get("text"):
                continue
            if btn.get("url"):
                buttons.append(
                    InlineKeyboardButton(str(btn["text"]), url=str(btn["url"]))
                )
            elif btn.get("callback_data"):
                buttons.append(
                    InlineKeyboardButton(
                        str(btn["text"]),
                        callback_data=str(btn["callback_data"])[:64],
                    )
                )
        if buttons:
            markup_rows.append(buttons)
    if not markup_rows:
        return None
    return InlineKeyboardMarkup(markup_rows)

File: internal/channel/telegram/keyboards.go
package telegram

import (
	"github.com/go-telegram/bot/models"

	"github.com/sahiy-backend/sahiy-agent/internal/domain/shared"
)

func languageInlineKeyboard() *models.InlineKeyboardMarkup {
	return &models.InlineKeyboardMarkup{
		InlineKeyboard: [][]models.InlineKeyboardButton{
			{
				{Text: "🇺🇿 O'zbek", CallbackData: "lang_uz"},
				{Text: "🇷🇺 Русский", CallbackData: "lang_ru"},
			},
			{
				{Text: "🇬🇧 English", CallbackData: "lang_en"},
				{Text: "🇨🇳 中文", CallbackData: "lang_zh"},
			},
		},
	}
}

func ratingInlineKeyboard() *models.InlineKeyboardMarkup {
	return &models.InlineKeyboardMarkup{
		InlineKeyboard: [][]models.InlineKeyboardButton{
			{
				{Text: "⭐", CallbackData: "rate_1"},
				{Text: "⭐⭐", CallbackData: "rate_2"},
				{Text: "⭐⭐⭐", CallbackData: "rate_3"},
				{Text: "⭐⭐⭐⭐", CallbackData: "rate_4"},
				{Text: "⭐⭐⭐⭐⭐", CallbackData: "rate_5"},
			},
		},
	}
}

func phoneRequestKeyboard(lang shared.Language) models.ReplyMarkup {
	return &models.ReplyKeyboardMarkup{
		Keyboard: [][]models.KeyboardButton{
			{{Text: phoneButtonLabel(lang), RequestContact: true}},
		},
		ResizeKeyboard:  true,
		OneTimeKeyboard: true,
	}
}

func mainMenuKeyboard(lang shared.Language) models.ReplyMarkup {
	labels := labelsFor(lang)
	return &models.ReplyKeyboardMarkup{
		Keyboard: [][]models.KeyboardButton{
			{{Text: labels.callback}, {Text: labels.newChat}},
			{{Text: labels.language}, {Text: labels.help}},
			{{Text: labels.productSearch}},
		},
		ResizeKeyboard: true,
		IsPersistent:   true,
	}
}

type menuLabelSet struct {
	callback, newChat, language, help, productSearch string
}

func labelsFor(lang shared.Language) menuLabelSet {
	switch lang.Code() {
	case shared.LangRu.Code():
		return menuLabelSet{
			callback: "📞 Обратный звонок", newChat: "🔄 Новый чат",
			language: "🌐 Сменить язык", help: "❓ Помощь", productSearch: "🔍 Поиск товаров",
		}
	case shared.LangEn.Code():
		return menuLabelSet{
			callback: "📞 Request a call", newChat: "🔄 New chat",
			language: "🌐 Change language", help: "❓ Help", productSearch: "🔍 Search products",
		}
	case shared.LangCyr.Code():
		return menuLabelSet{
			callback: "📞 Қайта қўнғироқ", newChat: "🔄 Янги суҳбат",
			language: "🌐 Тилни ўзгартириш", help: "❓ Ёрдам", productSearch: "🔍 Маҳsulot qidirish",
		}
	case shared.LangZh.Code():
		return menuLabelSet{
			callback: "📞 回电请求", newChat: "🔄 新对话",
			language: "🌐 更改语言", help: "❓ 帮助", productSearch: "🔍 搜索商品",
		}
	default:
		return menuLabelSet{
			callback: "📞 Qayta qo'ng'iroq", newChat: "🔄 Yangi suhbat",
			language: "🌐 Tilni o'zgartirish", help: "❓ Yordam", productSearch: "🔍 Mahsulot qidirish",
		}
	}
}

func phoneButtonLabel(lang shared.Language) string {
	switch lang.Code() {
	case shared.LangRu.Code():
		return "📱 Отправить номер телефона"
	case shared.LangEn.Code():
		return "📱 Send phone number"
	case shared.LangCyr.Code():
		return "📱 Телефон рақамни юбориш"
	case shared.LangZh.Code():
		return "📱 发送电话号码"
	default:
		return "📱 Telefon raqamni yuborish"
	}
}

func matchMenuAction(text string, lang shared.Language) (string, bool) {
	labels := labelsFor(lang)
	switch text {
	case labels.newChat:
		return "new_chat", true
	case labels.language:
		return "language", true
	case labels.help:
		return "help", true
	case labels.callback:
		return "callback", true
	case labels.productSearch:
		return "product_search", true
	default:
		return "", false
	}
}
